base/abstractarray.jl

```diff
@@ -126,6 +126,34 @@ _length(A) = length(A)
 endof(a::AbstractArray) = length(a)
 first(a::AbstractArray) = a[first(eachindex(a))]
 
+function fixate_eltype{T<:AbstractArray}(::Type{T}, default::Type)
+    @_pure_meta
+    aa_type = T
+    while aa_type.name.primary != AbstractArray
+        aa_type = supertype(aa_type)
+    end
+    if isa(aa_type.parameters[1], TypeVar)
+        varname = aa_type.parameters[1].name
+        new_params = collect(T.parameters)
+        for i in eachindex(new_params)
+            p = new_params[i]
+            if isa(p, TypeVar) && p.name == varname
+                new_params[i] = default
+            end
+        end
+        return T.name.primary{new_params...}
+    else
+        return T
+    end
+end
```

Comment:
@vtjnash I'm not sure I got the point. Does
> This function is fundamentally invalid. `T.name.primary` isn't a meaningful type and its type parameters don't have any fixed relation to the type parameters of `AbtractArray`

only relate to @pabloferz's proposed function, or also to the original `fixate_eltype` is this PR? Although the latter looks rather hacky, it seems to work quite well, so I'd like to understand any concern you might have.